src/mirage_flow.mli

```diff
@@ -50,63 +50,66 @@ module type S = sig
   (** [pp_write_error] is the pretty-printer for write errors. *)
 
   type flow
-  (** The type for flows. A flow represents the state of a single
-      reliable stream that is connected to an endpoint. *)
+  (** The type for flows. A flow represents the state of a single reliable
+      stream that is connected to an endpoint. *)
 
   val read: flow -> (Cstruct.t or_eof, error) result Lwt.t
-  (** [read flow] blocks until some data is available and returns a
-      fresh buffer containing it.
+  (** [read flow] blocks until some data is available and returns a fresh buffer
+      containing it.
 
       The returned buffer will be of a size convenient to the flow
       implementation, but will always have at least 1 byte.
 
-      If the remote endpoint calls [close] then calls to [read] will
-      keep returning data until all the in-flight data has been read.
-      [read flow] will return [`Eof] when the remote endpoint has
-      called [close] and when there is no more in-flight data.
-   *)
+      When [read] returns [`Eof] or an error, [close] (or [shutdown]) should be
+      called on the [flow] by the client. Once [read] returned [`Eof] or an
+      error, no subsequent [read] call will be successful. *)
 
   val write: flow -> Cstruct.t -> (unit, write_error) result Lwt.t
-  (** [write flow buffer] writes a buffer to the flow. There is no
-      indication when the buffer has actually been read and, therefore,
-      it must not be reused.  The contents may be transmitted in
-      separate packets, depending on the underlying transport. The
-      result [Ok ()] indicates success, [Error `Closed] indicates that the
-      connection is now closed and therefore the data could not be
-      written.  Other errors are possible. *)
+  (** [write flow buffer] writes a buffer to the flow. There is no indication
+      when the buffer has actually been sent and, therefore, it must not be
+      reused. The contents may be transmitted in separate packets, depending on
+      the underlying transport. The result [Ok ()] indicates success,
+      [Error `Closed] indicates that the connection is now closed and therefore
+      the data could not be written. Other errors are possible.
+
+      The call to [write] blocks until the buffer has been accepted by the
+      implementation (if a partial write occured, [write] will wait until the
+      remainder of the buffer has been accepted by the implementation).
```

Comment:
Hmm, does the *call* block or the *lwt promise* block? I think it is the latter, but the phrasing suggests the former.